File: backend/app/api/task_runner.py
"""
TRINITI Task Runner API
Main API router for task execution with OpenDevin agent integration
"""

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Any, Optional, Dict
import time
import uuid
from datetime import datetime

# Import our existing memory system
from ...services.TRINITIMemorySystem import TRINITIMemorySystem

router = APIRouter(prefix="/api", tags=["task_execution"])

# Initialize memory system
memory_system = TRINITIMemorySystem()

class TaskRequest(BaseModel):
    command: str
    metadata: Optional[Dict[str, Any]] = None
    timeout: Optional[int] = 30000  # 30 seconds default

class TaskResponse(BaseModel):
    id: str
    success: bool
    result: Any
    execution_time: float
    timestamp: int
    metadata: Optional[Dict[str, Any]] = None
    error: Optional[str] = None

class TaskExecutionOptions(BaseModel):
    timeout: Optional[int] = 30000
    retry_count: Optional[int] = 3
    priority: Optional[str] = "medium"
    tags: Optional[list[str]] = None

@router.post("/execute", response_model=TaskResponse)
async def execute_task(req: TaskRequest):
    """
    Execute a task using the TRINITI task execution system
    """
    start_time = time.time()
    task_id = f"task_{int(time.time() * 1000)}_{uuid.uuid4().hex[:8]}"

    try:
        # Initialize memory system if needed
        if not memory_system.isInitialized:
            await memory_system.initialize()

        # Execute task using our existing task executor
        from ...backend_server import task_executor

        # Create task execution request
        from ...backend_server import TaskExecutionRequest
        task_request = TaskExecutionRequest(
            task=req.command,
            timestamp=int(time.time() * 1000)
        )

        # Execute the task
        result = await task_executor.execute_task(task_request)

        # Record in memory system
        await memory_system.recordTask(
            task=req.command,
            result=result.result,
            metadata={
                'success': result.success,
                'duration': result.executionTime,
                'taskId': task_id,
                'type': 'task_execution',
                'tags': ['api_execution'] + (req.metadata.get('tags', []) if req.metadata else []),
                'priority': 5
            }
        )

        execution_time = time.time() - start_time

        return TaskResponse(
            id=task_id,
            success=result.success,
            result=result.result,
            execution_time=execution_time,
            timestamp=int(time.time() * 1000),
            metadata=req.metadata,
            error=None if result.success else str(result.result.get('error', 'Unknown error'))
        )

    except Exception as e:
        execution_time = time.time() - start_time
        error_message = str(e)

        # Record error in memory system
        try:
            await memory_system.recordTask(
                task=req.command,
                result={'error': error_message},
                metadata={
                    'success': False,
                    'duration': execution_time * 1000,
                    'taskId': task_id,
                    'type': 'task_execution_error',
                    'tags': ['api_execution', 'error'],
                    'priority': 1,
                    'errors': [error_message]
                }
            )
        except Exception as mem_error:
            print(f"Failed to record error in memory: {mem_error}")

        return TaskResponse(
            id=task_id,
            success=False,
            result={'error': error_message},
            execution_time=execution_time,
            timestamp=int(time.time() * 1000),
            metadata=req.metadata,
            error=error_message
        )

@router.get("/tasks", response_model=Dict[str, Any])
async def get_task_history(limit: int = 50, offset: int = 0):
    """
    Get task execution history from memory system
    """
    try:
        if not memory_system.isInitialized:
            await memory_system.initialize()

        # Get recent tasks from memory system
        recent_tasks = memory_system.getRecentTasks(limit + offset)

        # Apply pagination
        paginated_tasks = recent_tasks[offset:offset + limit]

        # Get total count
        stats = memory_system.getStats()

        return {
            "tasks": [
                {
                    "id": task.id,
                    "task": task.task,
                    "result": task.result,
                    "timestamp": task.timestamp,
                    "metadata": task.metadata
                }
                for task in paginated_tasks
            ],
            "total": stats.totalMemories,
            "limit": limit,
            "offset": offset
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve task history: {str(e)}")

@router.get("/tasks/stats", response_model=Dict[str, Any])
async def get_task_statistics():
    """
    Get task execution statistics from memory system
    """
    try:
        if not memory_system.isInitialized:
            await memory_system.initialize()

        stats = memory_system.getStats()

        return {
            "totalTasks": stats.totalMemories,
            "successfulTasks": stats.successfulTasks,
            "failedTasks": stats.failedTasks,
            "successRate": stats.successRate,
            "averageExecutionTime": stats.averageExecutionTime,
            "mostCommonTags": stats.mostCommonTags
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve statistics: {str(e)}")

@router.delete("/tasks")
async def clear_task_history():
    """
    Clear all task history from memory system
    """
    try:
        if not memory_system.isInitialized:
            await memory_system.initialize()

        await memory_system.clearAll()

        return {"message": "Task history cleared successfully"}

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to clear task history: {str(e)}")

@router.get("/health")
async def health_check():
    """
    Health check endpoint for the task runner API
    """
    return {
        "status": "healthy",
        "service": "triniti-task-runner",
        "timestamp": datetime.now().isoformat(),
        "memory_system_initialized": memory_system.isInitialized
    }

File: frontend/src/api/taskRunner.ts
/**
 * TRINITI Task Runner API Client
 * Frontend API client for task execution with proper error handling and type safety
 */

import axios, { AxiosResponse } from 'axios';
import { runTask as runTaskNL, runRawCommand } from '../services/task_runner';
import { TaskResult } from '../types/task';

// API Configuration - Use Vite environment variables
const API_BASE_URL = import.meta.env.VITE_API_URL ||
                    import.meta.env.REACT_APP_API_URL ||
                    'http://localhost:3000';
const API_TIMEOUT = 30000; // 30 seconds

// Configure axios defaults
axios.defaults.timeout = API_TIMEOUT;
axios.defaults.headers.common['Content-Type'] = 'application/json';

export interface TaskRequest {
  command: string;
  metadata?: Record<string, any>;
  timeout?: number;
}

export interface TaskResponse {
  id: string;
  success: boolean;
  result: any;
  execution_time: number;
  timestamp: number;
  metadata?: Record<string, any>;
  error?: string;
}

export interface TaskHistoryResponse {
  tasks: Array<{
    id: string;
    task: string;
    result: any;
    timestamp: number;
    metadata: any;
  }>;
  total: number;
  limit: number;
  offset: number;
}

export interface TaskStatistics {
  totalTasks: number;
  successfulTasks: number;
  failedTasks: number;
  successRate: number;
  averageExecutionTime: number;
  mostCommonTags: Array<{ tag: string; count: number }>;
}

export interface ApiError {
  message: string;
  status: number;
  details?: any;
}

/**
 * Execute a task via natural language processing
 */
export async function runTask(taskRequest: TaskRequest): Promise<TaskResult> {
  try {
    console.debug('Executing task with NL processing:', taskRequest.command);
    return await runTaskNL(taskRequest.command);
  } catch (error: any) {
    console.error('Task execution failed:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'Task execution failed',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}

/**
 * Execute a raw command directly (bypassing NL processing)
 */
export async function runRawTask(taskRequest: TaskRequest): Promise<TaskResult> {
  try {
    console.debug('Executing raw command:', taskRequest.command);
    return await runRawCommand(taskRequest.command);
  } catch (error: any) {
    console.error('Raw task execution failed:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'Raw task execution failed',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}

/**
 * Get task execution history
 */
export async function getTaskHistory(limit: number = 50, offset: number = 0): Promise<TaskHistoryResponse> {
  try {
    const response: AxiosResponse<TaskHistoryResponse> = await axios.get(
      `${API_BASE_URL}/api/tasks`,
      {
        params: { limit, offset }
      }
    );

    return response.data;
  } catch (error: any) {
    console.error('Failed to get task history:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'Failed to get task history',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}

/**
 * Get task execution statistics
 */
export async function getTaskStatistics(): Promise<TaskStatistics> {
  try {
    const response: AxiosResponse<TaskStatistics> = await axios.get(
      `${API_BASE_URL}/api/tasks/stats`
    );

    return response.data;
  } catch (error: any) {
    console.error('Failed to get task statistics:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'Failed to get task statistics',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}

/**
 * Clear all task history
 */
export async function clearTaskHistory(): Promise<{ message: string }> {
  try {
    const response: AxiosResponse<{ message: string }> = await axios.delete(
      `${API_BASE_URL}/api/tasks`
    );

    return response.data;
  } catch (error: any) {
    console.error('Failed to clear task history:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'Failed to clear task history',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}

/**
 * Check API health status
 */
export async function checkApiHealth(): Promise<{ status: string; service: string; timestamp: string }> {
  try {
    const response: AxiosResponse<{ status: string; service: string; timestamp: string }> = await axios.get(
      `${API_BASE_URL}/api/health`
    );

    return response.data;
  } catch (error: any) {
    console.error('API health check failed:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'API health check failed',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}

/**
 * Get API status information
 */
export async function getApiStatus(): Promise<{ backend: string; frontend: string; ready: boolean; task_runner: string }> {
  try {
    const response: AxiosResponse<{ backend: string; frontend: string; ready: boolean; task_runner: string }> = await axios.get(
      `${API_BASE_URL}/api/status`
    );

    return response.data;
  } catch (error: any) {
    console.error('Failed to get API status:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'Failed to get API status',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}

/**
 * Utility function to format execution time
 */
export function formatExecutionTime(time: number): string {
  if (time < 1000) {
    return `${Math.round(time)}ms`;
  } else {
    return `${(time / 1000).toFixed(2)}s`;
  }
}

/**
 * Utility function to check if API is available
 */
export async function isApiAvailable(): Promise<boolean> {
  try {
    await checkApiHealth();
    return true;
  } catch (error) {
    return false;
  }
}

// TRINITI FileOps Integration - Drop 77
export interface FileOperationRequest {
  operation: "create" | "read" | "delete";
  path: string;
  content?: string;
}

export interface FileOperationResponse {
  success: boolean;
  output: string;
}

/**
 * Execute file operations: create, read, delete
 */
export async function runFileOperation(
  operation: "create" | "read" | "delete",
  path: string,
  content: string = ""
): Promise<FileOperationResponse> {
  try {
    const response: AxiosResponse<FileOperationResponse> = await axios.post(
      `${API_BASE_URL}/api/file-ops`,
      {
        operation,
        path,
        content
      }
    );

    return response.data;
  } catch (error: any) {
    console.error('File operation failed:', error);

    const apiError: ApiError = {
      message: error.response?.data?.detail || error.message || 'File operation failed',
      status: error.response?.status || 500,
      details: error.response?.data
    };

    throw apiError;
  }
}
